modules/ghlint-rules/src/main/kotlin/net/twisterrob/ghlint/rules/InvalidExpressionUsageRule.kt

```diff
@@ -0,0 +1,82 @@
+package net.twisterrob.ghlint.rules
+
+import net.twisterrob.ghlint.model.ActionStep
+import net.twisterrob.ghlint.model.WorkflowStep
+import net.twisterrob.ghlint.rule.Example
+import net.twisterrob.ghlint.rule.Issue
+import net.twisterrob.ghlint.rule.Reporting
+import net.twisterrob.ghlint.rule.report
+import net.twisterrob.ghlint.rule.visitor.ActionVisitor
+import net.twisterrob.ghlint.rule.visitor.VisitorRule
+import net.twisterrob.ghlint.rule.visitor.WorkflowVisitor
+
+public class InvalidExpressionUsageRule : VisitorRule, ActionVisitor, WorkflowVisitor {
+
+	override val issues: List<Issue> = listOf(InvalidExpression)
+
+	override fun visitWorkflowUsesStep(reporting: Reporting, step: WorkflowStep.Uses) {
+		super.visitWorkflowUsesStep(reporting, step)
+
+		if (step.uses.uses.containsGitHubExpression()) {
+			reporting.report(InvalidExpression, step) {
+				"${it} contains a GitHub expression in the `uses` field."
+			}
+		}
+	}
+
+	override fun visitActionUsesStep(reporting: Reporting, step: ActionStep.Uses) {
+		super.visitActionUsesStep(reporting, step)
+
+		if (step.uses.uses.containsGitHubExpression()) {
```

Comment:
Looks like there's no common type between `ActionStep.Uses` and `WorkflowStep.Uses` that also conforms to `Component`. So couldn't generalise this logic.

Reply:
I just did the same, it's not the prettiest, but it works :)

https://github.com/TWiStErRob/net.twisterrob.ghlint/commit/948c28e3c68388d708a76c30aca0de349781089e#diff-4a935fa34331c8adc2d7846540f52162e56b0d8e0b296f074a8df4b2f9cc7b6fR29

Reply:
Btw, there are other options: EnvironmentFileOverwriteRule and EmptyEnvRule use 2 different approaches, which is a bit less hacky